catalyst/callbacks/profiler.py

```diff
@@ -0,0 +1,85 @@
+    Args:
+        profiler (torch.profiler.profile): The instantiated PyTorch profiler object
+    
+    Example:
+        .. code-block:: python
+            import torch  # noqa
+            from torch import nn, optim  # noqa
+            from torch.utils.data import DataLoader  # noqa
+            from catalyst import dl  # noqa
+            from catalyst.data.transforms import ToTensor  # noqa
+            from catalyst.contrib.datasets import MNIST  # noqa
+            from catalyst.callbacks.batch_overfit import BatchOverfitCallback  # noqa
+            from catalyst.callbacks.misc import TimerCallback  # noqa
+            from catalyst.utils.misc import set_global_seed  # noqa
+            from catalyst.callbacks.profiler import ProfilerCallback  # noqa
+            profiler = torch.profiler.profile(
+                profile_memory=True,
+                schedule=torch.profiler.schedule(
+                    wait=1,
+                    warmup=1,
+                    active=2))
+            with profiler:
+                torch.backends.cudnn.deterministic = True
+                torch.backends.cudnn.benchmark = True
+                torch.backends.cudnn.enabled = True
+                set_global_seed(42)
+                device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+                model = nn.Sequential(nn.Flatten(), nn.Linear(28 * 28, 10)).to(device)
+                criterion = nn.CrossEntropyLoss()
+                optimizer = optim.Adam(model.parameters(), lr=0.02)
+                loaders = {
+                    "train": DataLoader(
+                        MNIST('./', train=True, download=True, transform=ToTensor()),
+                        batch_size=256, num_workers=0, pin_memory=True),
+                    "valid": DataLoader(
+                        MNIST('./', train=False, download=True, transform=ToTensor()),
+                        batch_size=256, num_workers=0, pin_memory=True),
+                }
+                runner = dl.SupervisedRunner()
+                # model training
+                runner.train(
+                    model=model,
+                    criterion=criterion,
+                    optimizer=optimizer,
+                    loaders=loaders,
+                    num_epochs=1,
+                    logdir="./logs",
+                    valid_loader="valid",
+                    valid_metric="loss",
+                    minimize_valid_metric=True,
+                    verbose=True,
+                    callbacks=[ProfilerCallback(profiler)])
+                print(profiler.key_averages().table(sort_by="cpu_time_total", row_limit=10))
+                print(profiler.key_averages().table(sort_by="self_cpu_memory_usage", row_limit=10))
+                if torch.cuda.is_available():
+                    print(profiler.key_averages().table(sort_by="cuda_time_total", row_limit=10))
+    """
+
+    def __init__(self, profiler=None):
```

Comment:
let's also add an option to create `profiler` by params

Reply:
I mean, it would be also great to get ProfilerCallback example with Config API... something like
Config API - minimal example in https://github.com/catalyst-team/catalyst#minimal-examples

Reply:
If we create the `profiler` using params, we won't be able to wrap the training code in a `with profiler:` statement, or we would have to return the initialized profiler, but that seems over the top since the user could just instantiate the profiler by themselves in that case.

Reply:
you could use `self.profiler.__enter__()` to emulate with statement, something like:
```python
class ProfilerCallback:
  init(self, profiler=None, mode=loader or epoch): 
    profiler = profiler or init
    mode = mode

  on_epoch_start: profiler.__enter__() if mode == epoch else nothing
  on_loader_start: profiler.__enter__() if mode == loader else nothing
  on_loader_end: profiler.__exit__() if mode == loader else nothing
  on_epoch_end: profiler.__exit__() if mode == epoch else nothing
  ```

Reply:
I tried something similar after our discussion in the Catalyst slack, the profiler should be before the dataset instantiation which is at the `on_stage_start()` level

Reply:
you could add mode=stage, is there any problem with that?

Reply:
I think the problem was that the dataset instantiation happens before the callbacks are executed, so memory profiling would issue lots of warnings